ui/pages/confirmations/hooks/useCurrentConfirmation.ts

```diff
@@ -39,12 +39,24 @@ const useCurrentConfirmation = () => {
     let pendingConfirmation: Approval | undefined;
 
     if (paramsTransactionId) {
-      if (paramsTransactionId === currentConfirmation?.id) {
-        return;
-      }
+      // if gas prediction has changed, change the current confirmation
       pendingConfirmation = pendingConfirmations.find(
         ({ id: confirmId }) => confirmId === paramsTransactionId,
       );
+      const unconfirmedTransaction =
+        pendingConfirmation && unconfirmedTransactions[pendingConfirmation.id];
+      if (paramsTransactionId === currentConfirmation?.id) {
+        const gasHasChanged =
+          unconfirmedTransaction?.txParams &&
+          currentConfirmation?.txParams &&
+          JSON.stringify(unconfirmedTransaction.txParams) !==
+            JSON.stringify(currentConfirmation.txParams);
+        if (gasHasChanged) {
+          setCurrentConfirmation(unconfirmedTransaction);
```

Comment:
My high level concern with this file is we start with the pending approvals which try and abstract all confirmations into a consistent format, but we're progressively adding more coupling to specific domain and specific controller state.

This highlights a big issue with the current model in which a confirmation and all its data is static once it's put into redux hence this logic to update it based on the controller state.

Would a cleaner solution be for the confirmation we store in Redux just be an ID, type, and a data ID that corresponds to some metadata such as a transaction ID or signature ID.

Then the confirmation can remain static, but the transaction components we're adding can query the latest transaction state dynamically using that data ID / transaction ID.

This is very similar to the exiting pending approvals in which the transaction approval requests just include the `txId` in the `requestData`.

Reply:
Hey @matthewwalsh0 : what we are keeping in redux here is transient UI state and we get detail object from controller are put it here, to avoid doing that again and again in each component.

Reply:
Good point, in that case I see the benefits of copying it to Redux to avoid querying the controllers in the other components, but we still shouldn't need this manual JSON comparison to update the confirmation when the metadata changes.

I've simplified this hook in the transaction alerts PR by using just a `useMemo` to cache the confirmation, plus relying on deep equal selectors to ensure we automatically update the confirmation if the underlying transaction, signature, or approval changes.

https://github.com/MetaMask/metamask-extension/blob/feat/redesign-transaction-alerts/ui/pages/confirmations/hooks/useCurrentConfirmation.ts